### PR TITLE
Port to CE3

### PR DESCRIPTION
**Pull request by cquiroz**
This is an attempt to port clue to CE3
I don't know enough about CE3 to be sure the port is perfect but it could be a basis for testing.

### CODE REVIEW

File: core/shared/src/main/scala/clue/ApolloClient.scala [hunk 1]
```diff
@@ -643,7 +657,8 @@ class ApolloClient[F[_], S, CP, CE](
           val stream = bracket.flatMap(_ =>
             (
               Stream.eval(sendStart) >>
-                emitter.queue.dequeue
+                Stream
```
Comment (cquiroz) [inline core/shared/src/main/scala/clue/ApolloClient.scala:660]: `Queue` is not in fs2 anymore and not a stream. The method `fromQueueUnterminated` seems to do the conversion

File: scalajs/src/main/scala/clue/js/WebSocketJSBackend.scala [hunk 2]
```diff
@@ -16,77 +16,78 @@ import org.scalajs.dom.raw.Event
...
-            (
+            val open: F[Unit] = (
               for {
                 _ <- isOpen.set(true)
                 _ <- s"WebSocket open for URI [$uri]".traceF
               } yield cb(new WebSocketJSConnection(ws).asRight)
             ).uncancelable
-              .runAsync(_ => IO.unit)
-              .unsafeRunSync()
+            dispatcher.unsafeRunAndForget(open)
```
Comment (cquiroz) [inline scalajs/src/main/scala/clue/js/WebSocketJSBackend.scala:48]: Using `dispatcher` is the recommended alternative to run effects when interacting with a foreign interface.
Still it would be good to test it for real but I guess that needs to wait for explore to be ported

File: gen/rules/src/main/scala/clue/gen/GraphQLGen.scala [hunk 3]
```diff
@@ -8,6 +8,7 @@ import scala.meta._
 import cats.syntax.all._
 import metaconfig.Configured
 import cats.effect.IO
+import cats.effect.unsafe.implicits.global
```
Comment (cquiroz) [inline gen/rules/src/main/scala/clue/gen/GraphQLGen.scala:11]: This is executed as part of the scalafix, so we want to use the unsafe calls

File: core/shared/src/main/scala/clue/ApolloClient.scala [hunk 4]
```diff
@@ -78,16 +77,16 @@ class ApolloClient[F[_], S, CP, CE](
   state:                Ref[F, State[F, CP]],
   connectionStatus:     SignallingRef[F, PersistentClientStatus]
 )(implicit
-  F:                    ConcurrentEffect[F],
+  F:                    Async[F],
   backend:              PersistentBackend[F, CP, CE],
-  timer:                Timer[F],
+  timer:                Temporal[F],
```
Comment (rpiaggio) [inline core/shared/src/main/scala/clue/ApolloClient.scala:82]: It seems `Async` already extends `Temporal`. We don't want the `timer` parameter then.

File: scalajs/src/main/scala/clue/js/WebSocketJSBackend.scala (hunk 2 above)
Comment (rpiaggio) [reply scalajs/src/main/scala/clue/js/WebSocketJSBackend.scala:48]: I see. Maybe we can have the `Dispatcher.use` call here instead of passing it from the caller. It seems this would simplify the `ApolloClient` logic and we can remove the `dispatcher` parameter.

Quoting the docs:
`An instance of Dispatcher is very cheap - it allocates a single fiber so it is encouraged to instantiate it where necessary rather than wiring a single instance throughout an application.`

File: gen/rules/src/main/scala/clue/gen/GraphQLGen.scala (hunk 3 above)
Comment (rpiaggio) [inline gen/rules/src/main/scala/clue/gen/GraphQLGen.scala:11]: Why is this import needed?

File: scalajs/src/main/scala/clue/js/WebSocketJSBackend.scala (hunk 2 above)
Comment (cquiroz) [reply scalajs/src/main/scala/clue/js/WebSocketJSBackend.scala:48]: Right it is really up to use whether we create locally or pass it along. In this case I thought we could pass it at this level and create it at `ApolloClient`. I'm not sure of the pros/cons

File: core/shared/src/main/scala/clue/ApolloClient.scala (hunk 4 above)
Comment (cquiroz) [reply core/shared/src/main/scala/clue/ApolloClient.scala:82]: Good catch

File: gen/rules/src/main/scala/clue/gen/GraphQLGen.scala (hunk 3 above)
Comment (cquiroz) [reply gen/rules/src/main/scala/clue/gen/GraphQLGen.scala:11]: This files makes calls to `unsafeRunSync`. In general we may prefer to use dispatcher and let the runtime call it
But in this case this will run (If I understood correctly) in `scalafix` thus we should call `unsafe` methods directly

File: scalajs/src/main/scala/clue/js/WebSocketJSBackend.scala (hunk 2 above)
Comment (cquiroz) [reply scalajs/src/main/scala/clue/js/WebSocketJSBackend.scala:48]: If we use `Resource.use` that will produce a `F[A]` rather than `Unit` and that has to be executed by the `onopen` method. I think we have to use the `unsafe` methods here

Comment (rpiaggio) [reply scalajs/src/main/scala/clue/js/WebSocketJSBackend.scala:48]: I agree that we have to call `unsafe` here.

Can we create the `Dispatcher` instance here? I see as a pro that the `WebSoketBackend` API, as well as the `ApolloClient`, will both be kept pure, without the need of them to know about the `Dispatcher` at all.

File: gen/rules/src/main/scala/clue/gen/GraphQLGen.scala (hunk 3 above)
Comment (rpiaggio) [reply gen/rules/src/main/scala/clue/gen/GraphQLGen.scala:11]: I see. This provides an implicit `IORuntime` :+1:

File: scalajs/src/main/scala/clue/js/WebSocketJSBackend.scala (hunk 2 above)
Comment (cquiroz) [reply scalajs/src/main/scala/clue/js/WebSocketJSBackend.scala:48]: Ok, let's try that.

Comment (cquiroz) [reply scalajs/src/main/scala/clue/js/WebSocketJSBackend.scala]: Ok done

**Review by rpiaggio — APPROVED** [review]
LGTM. Let's try this!

**Comment by cquiroz** [comment]
Ok, let's discuss later how to do releases. Maybe we merge this at the same time than `crystal` to test everything together

**Comment by rpiaggio** [comment]
Sounds good.
